cli/cmd/install.go

```diff
@@ -200,66 +199,127 @@ func newInstallIdentityOptionsWithDefaults() *installIdentityOptions {
-	options := newInstallOptionsWithDefaults()
+// newCmdInstallConfig is a subcommand for `linkerd install config`
+func newCmdInstallConfig(options *installOptions) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "config [flags]",
+		Args:  cobra.NoArgs,
+		Short: "Output Kubernetes cluster-wide resources to install Linkerd",
+		Long: `Output Kubernetes cluster-wide resources to install Linkerd.
+
+This command provides Kubernetes configs necessary to install cluster-wide
+resources for the Linkerd control plane. This command should be followed by
+"linkerd install control-plane".`,
+		Example: `  # Default install.
+  linkerd install config | kubectl apply -f -
+
+  # Install Linkerd into a non-default namespace.
+  linkerd install config -l linkerdtest | kubectl apply -f -`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return installRunE(options, configStage, nil)
+		},
+	}
 
+	return cmd
+}
+
+// newCmdInstallControlPlane is a subcommand for `linkerd install control-plane`
+func newCmdInstallControlPlane(options *installOptions) *cobra.Command {
 	// The base flags are recorded separately so that they can be serialized into
 	// the configuration in validateAndBuild.
 	flags := options.recordableFlagSet()
 	installOnlyFlags := options.installOnlyFlagSet()
 
 	cmd := &cobra.Command{
-		Use:       fmt.Sprintf("install [%s|%s] [flags]", configStage, controlPlaneStage),
-		Args:      cobra.OnlyValidArgs,
-		ValidArgs: []string{configStage, controlPlaneStage},
-		Short:     "Output Kubernetes configs to install Linkerd",
-		Long: `Output Kubernetes configs to install Linkerd.
+		Use:   "control-plane [flags]",
+		Args:  cobra.NoArgs,
+		Short: "Output Kubernetes control plane resources to install Linkerd",
+		Long: `Output Kubernetes control plane resources to install Linkerd.
 
 This command provides Kubernetes configs necessary to install the Linkerd
-control-plane.`,
+control plane. It should be run after "linkerd install config".`,
 		Example: `  # Default install.
-  linkerd install | kubectl apply -f -
-
-  # Installation may also be broken up into two stages, by user privilege.
-  # First stage requires cluster-level privileges.
-  linkerd install config | kubectl apply -f -
-  # Second stage requires namespace-level privileges.
   linkerd install control-plane | kubectl apply -f -
 
   # Install Linkerd into a non-default namespace.
-  linkerd install config -l linkerdtest | kubectl apply -f -
   linkerd install control-plane -l linkerdtest | kubectl apply -f -`,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			stage, err := validateArgs(args, flags, installOnlyFlags)
-			if err != nil {
-				return err
-			}
-			if !options.ignoreCluster {
-				// TODO: consider cobra.SilenceUsage, so we can return errors from
-				// `RunE`, rather than calling `os.Exit(1)`
-				exitIfClusterExists()
-			}
-			if !options.skipChecks && stage == controlPlaneStage {
+			if !options.skipChecks {
```

Comment:
@ihcsim Thanks for all this testing!

Re: the `install` parent flags getting pushed up into the global flags when using a subcommand, yeah, it's how cobra works. i agree with you that it would be preferable to be more clear on what constitutes a truly global flag vs. a flag that is specific to `install` and its subcommands. i looked at repeating `cmd.Flags().AddFlagSet()` in the subcommands as you mentioned, but i think the current implementation expresses the functionality more explicitly, as `install` parent flags that are shared with subcommands.

re: `--skip-checks` vs. `--ignore-cluster`, i agree it's a bit confusing. have a look at https://github.com/linkerd/linkerd2/pull/2719#issuecomment-485894308 for the motivation. /cc @adleong 

`--skip-checks` is specific to the `linkerd install control-plane` subcommand, which requires a control plane namespace to exist to successfully be applied via kubectl.
`--ignore-cluster` applies to `install` and all subcommands, and is meant to override an existing control plane installation.

re:
```bash
# this serves no purposes, even though the flag is available
# the command still fails
$  bin/linkerd install control-plane --ignore-cluster
```
... that command succeeds if the control plane namespace exists, similar in behavior to `linkerd install`.

this discussion probably implies we could at least do a better job documenting these commands in the help output. let me know what you think.

Reply:
Sorry for jumping into this discussion after this has already been merged, but I just want to make sure I understand correctly what we landed on here.

I would expect that `--skip-checks` and `--ignore-cluster` are both flags on `install control-plane` that do different things:
* `--skip-checks` skips the check that the controller namespace exists which can be useful for grabbing the control plane yaml even if your cluster is not ready for it to be applied.  I would also expect this to skip other checks that verify that `install config` ran successfully like checking that CRDs, ClusterRoles, and ClusterRoleBindings were created.
* `--ignore-cluster` disregards the contents of the linkerd config configmap and can be used to completely override an existing control-plane instead of preserving existing config settings.

Furthermore, I would expect these flags to also be available on `linkerd install` as well since `linkerd install` includes `linkerd install control-plane`.

I would not expect either of those flags to be available on `linkerd install config`.  What do they do for that command?

Reply:
@adleong All permutations of subcommands and flags:

```
linkerd install --ignore-cluster
linkerd install config --ignore-cluster
linkerd install control-plane --ignore-cluster

# don't check for namespace existence (adding RBAC and CRD checks is a good idea)
linkerd install control-plane --skip-checks 
```

Having --ignore-cluster available on install and all subcommands makes consistent the idea that "linkerd install" doesn't work after installation. This makes the user experience consistent post-installation:

```bash
$ bin/go-run cli install [config|control-plane]
Linkerd has already been installed on your cluster in the linkerd namespace. Please run upgrade if you'd like to update this installation. Otherwise, use the --ignore-cluster flag.
```

I'm not sure of the use case for `linkerd install --skip-checks`. That flag checks for namespace existence, and fails if the namespace (and perhaps eventually CRDs, etc) doesn't exist. I think that's specific to `linkerd install control-plane`.

Reply:
Gotcha, this all makes sense to me.  Thanks for clarifying!

Reply:
Filed #2782 to track followup RBAC/CRD checks.